src/init/starship.zsh

```diff
@@ -3,93 +3,90 @@
+    __starship_time() {
+        echo $(( int(rint(EPOCHREALTIME * 1000)) ))
+    }
+else
+    __starship_time() {
+        ::STARSHIP:: time
     }
 fi
 
 # Will be run before every prompt draw
-starship_precmd() {
+__starship_precmd() {
     # Save the status, because commands in this pipeline will change $?
     STARSHIP_CMD_STATUS=$?
 
     # Compute cmd_duration, if we have a time to consume, otherwise clear the
     # previous duration
-    if (( ${+STARSHIP_START_TIME} )); then
-        __starship_get_time && (( STARSHIP_DURATION = STARSHIP_CAPTURED_TIME - STARSHIP_START_TIME ))
-        unset STARSHIP_START_TIME
+    if (( ${+STARSHIP_CMD_START_TIME} )); then
+        STARSHIP_CMD_DURATION=$(( $(__starship_time) - STARSHIP_CMD_START_TIME ))
+        unset STARSHIP_CMD_START_TIME
     else
-        unset STARSHIP_DURATION
+        unset STARSHIP_CMD_DURATION
     fi
 
     # Use length of jobstates array as number of jobs. Expansion fails inside
     # quotes so we set it here and then use the value later on.
     STARSHIP_JOBS_COUNT=${#jobstates}
 }
-starship_preexec() {
-    __starship_get_time && STARSHIP_START_TIME=$STARSHIP_CAPTURED_TIME
+__starship_preexec() {
+    STARSHIP_CMD_START_TIME="$(__starship_time)"
 }
 
 # If precmd/preexec arrays are not already set, set them. If we don't do this,
-# the code to detect whether starship_precmd is already in precmd_functions will
-# fail because the array doesn't exist (and same for starship_preexec)
+# the code to detect whether __starship_precmd is already in precmd_functions will
+# fail because the array doesn't exist (and same for __starship_preexec)
 (( ! ${+precmd_functions} )) && precmd_functions=()
 (( ! ${+preexec_functions} )) && preexec_functions=()
 
 # If starship precmd/preexec functions are already hooked, don't double-hook them
 # to avoid unnecessary performance degradation in nested shells
-if [[ -z ${precmd_functions[(re)starship_precmd]} ]]; then
-    precmd_functions+=(starship_precmd)
-fi
-if [[ -z ${preexec_function[(re)starship_preexec]} ]]; then
-    preexec_functions+=(starship_preexec)
-fi
+(( ! ${precmd_functions[(Ie)__starship_precmd]} )) && precmd_functions+=(__starship_precmd)
+(( ! ${preexec_functions[(Ie)__starship_preexec]} )) && preexec_functions+=(__starship_preexec)
 
 # Set up a function to redraw the prompt if the user switches vi modes
-starship_zle-keymap-select() {
-    zle reset-prompt
-}
-
-## Check for existing keymap-select widget.
-local existing_keymap_select_fn=$widgets[zle-keymap-select];
-# zle-keymap-select is a special widget so it'll be "user:fnName" or nothing. Let's get fnName only.
-existing_keymap_select_fn=${existing_keymap_select_fn//user:};
-if [[ -z $existing_keymap_select_fn ]]; then
-    zle -N zle-keymap-select starship_zle-keymap-select;
+# zle-keymap-select is a special widget named like "user:handler-name"
+zle_keymap_select=${${widgets[zle-keymap-select]}//user:}
```

Comment:
consider prefixing it too: `__starship_preserved_zle_keymap_select`

Reply:
Well, this is very much a work in progress. I haven't even pushed all my local commits. FWIW, my latest iteration looks like this:
```sh
__zle_keymap_select=${widgets[zle-keymap-select]#user:}
```

I didn't choose `__starship` prefix because it has been used exclusively for things **added** by Starship itself.

Reply:
The reason recommended it because, `__zle_keymap_select` is too generic. Another tool can try to do the same thing and use the same variable name, and that could mess with starship's intialization.

But prefixing it with `__starship` brings down that possibility to almost zero.

Reply:
That's a good point! Honestly, I'm no fan of verbose identifiers but you are right it is better to be safe than sorry.